### PR TITLE
feat: add event emission and fee collection logic with simple tests

### PR DESCRIPTION
**Pull request by 0xIcarus**
## Description

Introduced events `WithdrawStakingRewardEvent` and `PoolLowBalanceEvent` to help with creating alerts for the governed gas pool to signal when funds are low and also reset when replenished. Extends pool accounting and entry points: treasury deposits are tracked, gas fees/treasury/deposits funnel into the pool consistently, and withdrawal of staking rewards happen with event emission.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [x] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [x] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

Unit tests added in aptos_framework::governed_gas_pool cover core flows like deposit paths, funding from pool, treasury accounting, event emission validation and edge cases like withdrawal asserts on insufficient pool balance and overflow guard in dynamic threshold calculation.

## Key Areas to Review

- Structs: `WithdrawStakingRewardEvent`, `PoolLowBalanceEvent`, `GovernedGasPoolExtension`, `ThresholdParams`
- Views/counters: `get_balance`, `get_treasury_deposited`
- Deposits/flows: `deposit`, `deposit_from`, `deposit_from_fungible_store`, `deposit_gas_fee`, `deposit_gas_fee_v2`, `deposit_treasury`

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by 0xmovses** [comment]
If this is ready for review want to take it out of review and add a description? Thanks!

File: aptos-move/framework/aptos-framework/sources/governed_gas_pool.move [hunk 1]
```diff
@@ -30,6 +32,17 @@ module aptos_framework::governed_gas_pool {
         amount: u64,
     }
 
+    /// Event emitted when pool balance goes below the configured low threshold
+    #[event]
+    struct PoolLowBalance has drop, store {
```
Comment (musitdev) [inline aptos-move/framework/aptos-framework/sources/governed_gas_pool.move]: The idiomatic way to name an event is to end with Event.

File: aptos-move/framework/aptos-framework/sources/governed_gas_pool.move [hunk 2]
```diff
@@ -41,6 +54,23 @@ module aptos_framework::governed_gas_pool {
     struct GovernedGasPoolExtension has key {
         deposited_treasury_counter: u64,
         withdraw_staking_reward_events: EventHandle<WithdrawStakingRewardEvent>,
+        /// Event handle for low-balance alerts
+        low_balance_events: EventHandle<PoolLowBalance>,
+        /// Whether we've already emitted a low-balance alert for the current state
+        low_event_emitted: bool,
+    }
```
Comment (musitdev) [inline aptos-move/framework/aptos-framework/sources/governed_gas_pool.move:61]: The GovernedGasPoolExtension is going to be deployed on testnet with the staking update. If it's the case these struct changes will block the PR framework update. So I see 2 possibilities, merge this PR before the staking installation in testnet or use a new struct.
I'll prefer to merge this PR before the staking is deployed to simplify the code.
cc: @areshand @0xmovses

File: aptos-move/framework/aptos-framework/sources/governed_gas_pool.move [hunk 3]
```diff
@@ -120,6 +154,82 @@ module aptos_framework::governed_gas_pool {
...
+            0
+        } else {
+            let p = borrow_global<ThresholdParams>(@aptos_framework);
+            let base = (p.base_per_validator as u128);
+            let n = (p.num_validators_hint as u128);
+            let runway = (p.runway_epochs as u128);
+            let bps = (10000u128 + (p.safety_bps as u128));
+            let product = base * n * runway * bps;
+            let low_u128 = product / 10000u128;
+            assert!(low_u128 <= (18446744073709551615u128), 0);
```
Comment (musitdev) [inline aptos-move/framework/aptos-framework/sources/governed_gas_pool.move]: We should create a new error or use an existing one instead of returning 0.

File: aptos-move/framework/aptos-framework/sources/governed_gas_pool.move [hunk 4]
```diff
@@ -120,6 +154,82 @@ module aptos_framework::governed_gas_pool {
...
+        if (!exists<ThresholdParams>(@aptos_framework)) {
+            0
+        } else {
+            let p = borrow_global<ThresholdParams>(@aptos_framework);
+            let base = (p.base_per_validator as u128);
+            let n = (p.num_validators_hint as u128);
+            let runway = (p.runway_epochs as u128);
+            let bps = (10000u128 + (p.safety_bps as u128));
+            let product = base * n * runway * bps;
+            let low_u128 = product / 10000u128;
```
Comment (musitdev) [inline aptos-move/framework/aptos-framework/sources/governed_gas_pool.move:173]: Put the real equation or its description in comment to validate the implementation is correct.
Why 10000?

File: aptos-move/framework/aptos-framework/sources/governed_gas_pool.move [hunk 5]
```diff
@@ -120,6 +154,82 @@ module aptos_framework::governed_gas_pool {
...
+        }
+    }
+
+    #[view]
+    public fun get_dynamic_low_threshold(): u64 acquires ThresholdParams {
+        compute_dynamic_low_threshold()
+    }
+
+    /// Set threshold parameters
+    public fun set_threshold_params(
```
Comment (musitdev) [inline aptos-move/framework/aptos-framework/sources/governed_gas_pool.move]: We should define it as an entry function (like initialize_governed_gas_pool_extension) to be able to call it from the CLI to avoid creating a script?

File: aptos-move/framework/aptos-framework/sources/governed_gas_pool.move [hunk 6]
```diff
@@ -120,6 +154,82 @@ module aptos_framework::governed_gas_pool {
...
+    }
+
+    /// Emit a one-shot low-balance event when crossing below threshold
+    fun maybe_emit_low_balance_event() acquires GovernedGasPool, GovernedGasPoolExtension, ThresholdParams {
+        let pool_address = governed_gas_pool_address();
+        let total_balance = coin::balance<AptosCoin>(pool_address);
+
+        let dyn_low = compute_dynamic_low_threshold();
+        // If no dynamic params set, treat low threshold as 0 to avoid false alerts unless desired
+        let low_threshold = dyn_low;
```
Comment (musitdev) [inline aptos-move/framework/aptos-framework/sources/governed_gas_pool.move]: Why you rename dyn_low to low_threshold ?

File: aptos-move/framework/aptos-framework/sources/governed_gas_pool.move [hunk 7]
```diff
@@ -14,6 +14,8 @@ module aptos_framework::governed_gas_pool {
     use std::features;
     use aptos_framework::signer;
     use aptos_framework::aptos_account::Self;
+    
+    use aptos_framework::timestamp;
```
Comment (musitdev) [inline aptos-move/framework/aptos-framework/sources/governed_gas_pool.move:18]: timestamp is only use for testing. Add the #[test_only] annotation.

File: aptos-move/framework/aptos-framework/sources/governed_gas_pool.move [hunk 8]
```diff
@@ -487,4 +609,38 @@ module aptos_framework::governed_gas_pool {
...
+        mint_for_test(governed_gas_pool_address(), 2000);
+
+        // ensure flag is false initially
+        let ext0 = borrow_global<GovernedGasPoolExtension>(@aptos_framework);
+        assert!(!ext0.low_event_emitted, 1);
+
+        // withdraw 1500 -> pool = 500 < 1000 => should emit event and set flag
+        let c = withdraw_staking_reward<AptosCoin>(1500);
+        let ext1 = borrow_global<GovernedGasPoolExtension>(@aptos_framework);
+        assert!(ext1.low_event_emitted, 2);
```
Comment (musitdev) [inline aptos-move/framework/aptos-framework/sources/governed_gas_pool.move:643]: First, you should test a withdrawal with a threshold above to test the normal process. We must test that we don't generate a false event.

File: aptos-move/framework/aptos-framework/sources/governed_gas_pool.move [hunk 9]
```diff
@@ -487,4 +609,38 @@ module aptos_framework::governed_gas_pool {
...
+
+        // deposit back above threshold -> flag resets; move withdrawn coin to user
+        mint_for_test(governed_gas_pool_address(), 2000);
+        // re-evaluate threshold after direct mint (bypasses deposit path)
+        maybe_emit_low_balance_event();
+        coin::deposit(signer::address_of(user), c);
```
Comment (musitdev) [inline aptos-move/framework/aptos-framework/sources/governed_gas_pool.move]: Put this line at the end, it's only to make the compiler happy, it's not part of the test.

**Review by musitdev — COMMENTED** [review]
Some minor remarks. I think we should push this PR before the staking deployment.

**Comment by 0xIcarus** [comment]
Closing this PR as per suggestions from @areshand to move this logic offchain
